tools/reduce.py

```diff
@@ -212,9 +213,9 @@ def fix_path(maybe_path: str) -> str:
                 match = re.match(self.path_flag_pattern, arg)
```

Comment:
There's a few things that make my teeth itch about this:

- We handle the TU path explicitly. But that's just a path; why not handle it with fix_path?
- We hand random strings to fix_path on line 221 and then (maybe) overwrite them on the next iteration of the loop.

If fix_path can handle non-paths (it gives up on line 206), another way to look at this could be:

```py
flag_pattern = re.compile('(?P<prefix>(.*=)?)(?P<rest>.*)')
[m.group("prefix") + fix_path(m.group("rest")) for m in map(flag_pattern.match, self.entry.arguments)]
```

Maybe you don't need the file path pattern after all. Or you could elaborate flag_pattern to understand them as part of "prefix".

Reply:
> We handle the TU path explicitly. But that's just a path; why not handle it with fix_path?

Because the TU path needs different handling. Other paths need to be made absolute. The TU path needs to be made relative. I wrote this earlier in the doc comment for this function:

```
        Post-condition:
        - Include paths, such as using -I and -isysroot, use absolute paths
        - The path for the main TU has a relative path
```

> We hand random strings to fix_path on line 221 and then (maybe) overwrite them on the next iteration of the loop.

I think there's an argument to be made that we're not handing random strings to `fix_path` (at least, it's very unlikely that this happens). The reason that branch exists is because you can have argument lists which have either `["-Ipath/to/dir"]` (joined) or `["-I", "path/to/dir"]` (split), since many flags support both styles. Once we've parsed the flag, we're checking if the path is empty. If the path is empty, that must mean that the flag is using the split style, not the joined style. If it's using the split style, the next argument must be the path for the flag (unless the command-line itself was malformed, but it's reasonable to assume that's not the case).

Reply:
> Because the TU path needs different handling.

Right... sorry for being inattentive to that.

> I think there's an argument to be made that we're not handing random strings to fix_path (at least, it's very unlikely that this happens).

...Good point. Because the flag path pattern only names flags which expect paths. In which case maybe fix_path should be noisy if reaches that give-up case.

Reply:
I've added a logging statement for that case.